sdk/nodejs/eventhub/zMixins.ts

```diff
@@ -498,3 +501,180 @@ export class EventHubFunction extends appservice.Function<EventHubContext, strin
+     */
+    resourceGroupName: pulumi.Input<string>;
+
+    /**
+     * Azure Resource ID.
+     */
+    id: pulumi.Input<string>;
+}
+
+/**
+ * A callback-based subscription to events coming from Event Grid. Creates an Azure Function and
+ * an Event Grid Event Subscription with the webhook URL pointing to the Azure Function.
+ */
+export class EventGridCallbackSubscription<T> extends appservice.EventSubscription<EventGridContext<T>, EventGridEvent<T>, appservice.FunctionDefaultResponse> {
+    public readonly subscription: EventGridEventSubscription;
+
+    constructor(name: string,
+                scope: EventGridScope,
+                args: EventGridCallbackSubscriptionArgs<T>,
+                opts: pulumi.ComponentResourceOptions = {}) {
+
+        const { resourceGroupName, location } =
+            appservice.getResourceGroupNameAndLocation(args, pulumi.output(scope.resourceGroupName));
+
+        super("azure:eventhub:EventGridCallbackSubscription",
+              name,
+              new EventGridFunction(name, args),
+              { ...args, resourceGroupName, location },
+              opts);
+
+        const keys = pulumi.output(this.functionApp.getHostKeys());
+        const key = keys.apply(keys => keys.systemKeys["eventgrid_extension"]);
+        const url = pulumi.interpolate`https://${this.functionApp.defaultHostname}/runtime/webhooks/eventgrid?functionName=${name}&code=${key}`;
+
+        this.subscription = new EventGridEventSubscription(name, {
+            webhookEndpoint: { url },
+            scope: scope.id,
+            ...args,
+        }, { ...opts, parent: this });
+
+        this.registerOutputs();
+    }
+}
+
+type ResourceGroupEvent =
+    eventgrid.ResourceActionCancelData |
+    eventgrid.ResourceActionFailureData |
+    eventgrid.ResourceActionSuccessData |
+    eventgrid.ResourceDeleteCancelData |
+    eventgrid.ResourceDeleteFailureData |
+    eventgrid.ResourceDeleteSuccessData |
+    eventgrid.ResourceWriteCancelData |
+    eventgrid.ResourceWriteFailureData |
+    eventgrid.ResourceWriteSuccessData;
+
+// This can't be placed in core mixins because that would lead to circular references in core-appservice-eventhub
+declare module "../core/resourceGroup" {
+    interface ResourceGroup {
+        /**
+         * Creates a new subscription to events fired to the handler provided from Event Grid
+         * on any Resource event within the resource group.
+         */
+        onGridEvent(name: string,
+                    args: EventGridCallbackSubscriptionArgs<ResourceGroupEvent>,
+                    opts?: pulumi.ComponentResourceOptions,
+                   ): EventGridCallbackSubscription<ResourceGroupEvent>;
+    }
+}
+
+ResourceGroup.prototype.onGridEvent = function(this: ResourceGroup, name, args, opts) {
+    return new EventGridCallbackSubscription(
+        name, { id: this.id, resourceGroupName: this.name },
+        args, { parent: this, ...opts });
+};

```

Comment:
i'm on the fence about this.   How core do you think this functionality is?  Do people listen to resource group events a lot?  I ask since i think of ResourceGroup as a very core type, and it feels like a slightly strange one-off coupling between ResourceGroup and EventGrid.

Reply:
I have no idea, probably not too much. I've never used it before, but thought it's a decent use case. What's the alternative?

Reply:
as an example, for awsx we decided to invert this for things like metrics.  Instead of exposing off the resources, you can just go into each package and work with the metrics.  I somewhat feel like this should be gotten through the eventgrid module and you can say "i'd like the event grid events for these N different resource types".

Reply:
Can you give an example of a call that you have in mind? `EventGrid.onResouceGroup(resrouceGroup, {...})`?

Reply:
> Can you give an example of a call that you have in mind? `EventGrid.onResouceGroup(resrouceGroup, {...})?`

Right.  it could even be a top level module function.  like `eventGrid.onResourceGroupEvent`.  It might be in a sub-module.  For example, in awsx, we do this:

`awsx.cognito.metrics.compromisedCredentialsRisk(...)`.

So i'd be interested in exploring:

`azure.eventhub.onResourceGroupGridEvent(...)` or
`azure.eventhub.events.onResourceGroupdGridEvent(...)`.